// @lwrjs/router/build/cjs/modules/lwr/contextProvider/contextProvider.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/router/src/modules/lwr/contextProvider/contextProvider.ts
__markAsModule(exports);
__export(exports, {
  currentPageReferenceContextualizer: () => currentPageReferenceContextualizer,
  currentRouteContextualizer: () => currentRouteContextualizer,
  currentViewContextualizer: () => currentViewContextualizer,
  navigationContextContextualizer: () => navigationContextContextualizer,
  provideContext: () => provideContext
});
var import_lwc = __toModule(require("lwc"));
var import_navigation = __toModule(require("lwr/navigation"));
var currentRouteContextualizer = (0, import_lwc.createContextProvider)(import_navigation.CurrentRoute);
var currentPageReferenceContextualizer = (0, import_lwc.createContextProvider)(import_navigation.CurrentPageReference);
var currentViewContextualizer = (0, import_lwc.createContextProvider)(import_navigation.CurrentView);
var navigationContextContextualizer = (0, import_lwc.createContextProvider)(import_navigation.NavigationContext);
function provideContext(contextValue, providerNode, contextualizer, contextualAdapter) {
  contextualAdapter.setContext(providerNode, contextValue);
  contextualizer(providerNode, {
    consumerConnectedCallback: contextualAdapter.subscribeContext.bind(contextualAdapter, providerNode),
    consumerDisconnectedCallback: contextualAdapter.unsubscribeContext.bind(contextualAdapter, providerNode)
  });
}
